gimie/utils/text.py

```diff
@@ -0,0 +1,318 @@
+    ['this is', 'is a', 'a test']
+    """
+    return [
+        " ".join(tokens[i : i + size])
+        for i in range(0, len(tokens) - size + 1)
+    ]
+
+
+def get_ngram_counts(
+    doc: str, ngram_range: Tuple[int, int] = (1, 1)
+) -> Counter[str]:
+    """Get ngram counts for a document. The ngram range is inclusive.
+
+    Parameters
+    ----------
+    doc:
+        Document to extract ngrams from.
+    ngram_range:
+        Inclusive range of ngram sizes to extract.
+
+    Examples
+    --------
+    >>> get_ngram_counts("Red roses red.", ngram_range=(1, 2))
+    Counter({'red': 2, 'roses': 1, 'red roses': 1, 'roses red': 1})
+    """
+    ngram_counts: Counter[str] = Counter()
+    tokens = tokenize(doc)
+    for size in range(ngram_range[0], ngram_range[1] + 1):
+        ngram_counts += Counter(extract_ngrams(tokens, size))
+    return ngram_counts
+
+
+def normalize_csr_rows(X: sp.csr_matrix, norm: str = "l1") -> sp.csr_matrix:
+    """Normalize rows of a CSR matrix in place.
+
+    Parameters
+    ----------
+    X:
+        CSR matrix to normalize.
+    norm:
+        Norm to use for normalization. Either "l1" or "l2".
+
+    Examples
+    --------
+    >>> X = sp.csr_matrix([[1, 2], [3, 4]], dtype=np.float64)
+    >>> normalize_csr_rows(X, norm="l1").toarray()
+    array([[0.33333333, 0.66666667],
+           [0.42857143, 0.57142857]])
+    >>> normalize_csr_rows(X, norm="l2").toarray()
+    array([[0.4472136 , 0.89442719],
+           [0.6       , 0.8       ]])
+    """
+    norm_func = {
+        "l1": lambda x: np.abs(x).sum(),
+        "l2": lambda x: np.sqrt((x**2).sum()),
+    }[norm]
+
+    for i in range(X.shape[0]):
+        if X[i].sum() == 0.0:
+            continue
+
+        X[i, :] /= norm_func(X[i].data)
+    return X
+
+
+@dataclass
+class TfidfConfig:
+    """Configuration for TfidfVectorizer.
+
+    For more information on tf-idf, see https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfTransformer.html
+
+    Parameters
+    ----------
+    max_features:
+        Maximum number of features to keep. If None, all features are kept.
+    ngram_range:
+        Inclusive range of ngram sizes to extract.
+    smooth_idf:
+        Smooth idf weights by adding a constant 1 to the numerator and denominator
+        of the idf as if an extra document was seen containing every term once,
+        preventing zero divisions.
+    vocabulary:
+        Vocabulary to use. If None, the vocabulary is inferred from the data.
+    norm:
+        Normalization to use for the tfidf matrix. Either "l1" or "l2".
+    sublinear_tf:
+        Apply sublinear tf scaling, i.e. replace tf with 1 + log(tf).
+    """
+
+    max_features: Optional[int] = None
+    ngram_range: Tuple[int, int] = (1, 1)
+    smooth_idf: bool = True
+    vocabulary: Optional[Dict[str, int]] = None
+    norm: Optional[Literal["l1", "l2"]] = None
+    sublinear_tf: bool = False
+
+
+class TfidfVectorizer(BaseModel):
```

Comment:
This class is quite big. Is there a reason you didn't follow the split done in sklearn? (CountVectorizer vs TfidfTransformer)  

Count Vectoriser could encompass : 

```python
counts_records: List[Counter[str]] = [
            get_ngram_counts(doc, self.config.ngram_range) for doc in data
        ]
vocab = self.config.vocabulary or self._get_vocabulary(counts_records)
```